File: 03_BDD/01_Bowling/java/01_start/test/main/java/agilepartner/java/bowling/BowlingSteps.java
package agilepartner.java.bowling;

import cucumber.api.java.en.Given;
import cucumber.api.java.en.Then;
import cucumber.api.java.en.When;
import org.junit.Assert;

import java.util.Arrays;

public class BowlingSteps {

    private Game game;

    @Given("^a new bowling game$")
    public void givenANewBowlingGame() {
        game = new Game();
    }

    @When("^all of my balls are landing in the gutter$")
    public void whenAllOfMyBallsAreLandingInTheGutter() {
        for (int i = 0; i < 20; i++) {
            game.addRoll(0);
        }
    }

    @When("^all of my balls are strike$")
    public void all_of_my_balls_are_strike() throws Throwable {
        for (int i = 0; i < 21; i++) {
            game.addRoll(10);
        }
    }

    @Then("^my total score should be (\\d+)$")
    public void thenMyTotalScoreShouldBe(int score) {
        Assert.assertEquals(score, game.calculateScore());
    }

    @When("^I roll the following series:(.*)$")
    public void whenIRollTheFollowingSeries(String series) {
        Arrays.asList(series.trim().split(",")).forEach(e -> game.addRoll(Integer.valueOf(e)));
    }
}

File: 06_Bowling_BDD/ts/02_solution/test/bowling.steps.ts
import { DefineStepFunction, defineFeature, loadFeature } from "jest-cucumber";
import { Game } from "../src/bowling";

const feature = loadFeature("features/bowling.feature");

defineFeature(feature, (test) => {
  let game: Game;

  const givenANewGame = (given: DefineStepFunction) => {
    given("a new bowling game", () => {
      game = new Game();
    });
  };

  const thenMyScoreShouldBe = (then: DefineStepFunction) => {
    then(/^my total score should be (\d+)$/, (score) => {
      expect(game.score()).toBe(parseInt(score));
    });
  };

  const whenIRollTheFollowingSeries = (when: DefineStepFunction) => {
    when(/^I roll the following series: (.*)$/, (param) => {
      const pins = (param as string).trim().split(",");
      pins.map((p) => game.roll(parseInt(p)));
    });
  };

  test("Gutter game", ({ given, when, then }) => {
    givenANewGame(given);

    when("all of my balls land in the gutter", () => {
      for (let i = 0; i < 20; i++) {
        game.roll(0);
      }
    });

    thenMyScoreShouldBe(then);
  });

  test("Perfect game", ({ given, when, then }) => {
    givenANewGame(given);

    when("all of my balls are srikes", () => {
      for (let i = 0; i < 12; i++) {
        game.roll(10);
      }
    });

    thenMyScoreShouldBe(then);
  });

  test("All nines", ({ given, when, then }) => {
    givenANewGame(given);
    whenIRollTheFollowingSeries(when);
    thenMyScoreShouldBe(then);
  });

  test("All spares", ({ given, when, then }) => {
    givenANewGame(given);
    whenIRollTheFollowingSeries(when);
    thenMyScoreShouldBe(then);
  });

  test("Normal game", ({ given, when, then }) => {
    givenANewGame(given);
    whenIRollTheFollowingSeries(when);
    thenMyScoreShouldBe(then);
  });
});
